sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java

```diff
@@ -569,10 +613,353 @@ public boolean getValidate() {
+        }
+        return ImmutableList.copyOf(avroSources);
+      }
+    }
+
+    private static class TransformingSource<T, V> extends BoundedSource<V> {
+      private final BoundedSource<T> boundedSource;
+      private final SerializableFunction<T, V> function;
+      private final Coder<V> outputCoder;
+
+      public TransformingSource(
+          BoundedSource<T> boundedSource,
+          SerializableFunction<T, V> function,
+          Coder<V> outputCoder) {
+        this.boundedSource = boundedSource;
+        this.function = function;
+        this.outputCoder = outputCoder;
+      }
+
+      @Override
+      public List<? extends BoundedSource<V>> splitIntoBundles(
+          long desiredBundleSizeBytes, PipelineOptions options) throws Exception {
+        return Lists.transform(
+            boundedSource.splitIntoBundles(desiredBundleSizeBytes, options),
+            new Function<BoundedSource<T>, BoundedSource<V>>() {
+              @Override
+              public BoundedSource<V> apply(BoundedSource<T> input) {
+                return new TransformingSource<>(input, function, outputCoder);
+              }
+            });
+      }
+
+      @Override
+      public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+        return boundedSource.getEstimatedSizeBytes(options);
+      }
+
+      @Override
+      public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+        return boundedSource.producesSortedKeys(options);
+      }
+
+      @Override
+      public BoundedReader<V> createReader(PipelineOptions options) throws IOException {
+        return new TransformingReader(boundedSource.createReader(options));
+      }
+
+      @Override
+      public void validate() {
+        boundedSource.validate();
+      }
+
+      @Override
+      public Coder<V> getDefaultOutputCoder() {
+        return outputCoder;
+      }
+
+      private class TransformingReader extends BoundedReader<V> {
+        private final BoundedReader<T> boundedReader;
+
+        private TransformingReader(BoundedReader<T> boundedReader) {
+          this.boundedReader = boundedReader;
+        }
+
+        @Override
+        public synchronized BoundedSource<V> getCurrentSource() {
+          return TransformingSource.this;
+        }
+
+        @Override
+        public boolean start() throws IOException {
+          return boundedReader.start();
+        }
+
+        @Override
+        public boolean advance() throws IOException {
+          return boundedReader.advance();
+        }
+
+        @Override
+        public V getCurrent() throws NoSuchElementException {
+          T current = boundedReader.getCurrent();
+          return function.apply(current);
+        }
+
+        @Override
+        public synchronized BoundedSource<V> splitAtFraction(double fraction) {
+          return new TransformingSource<>(
+              boundedReader.splitAtFraction(fraction), function, outputCoder);
+        }
```

Comment:
can you verify that you implemented all the functions? Looks to me like there is a missing getCurrentFraction or something like that, which is key in Dataflow to triggering dynamic work rebalancing.